Add pydantic model

Speters1086 opened the pull request:
This PR introduces a Pydantic data model built off of the JSON schema. This is useful for making use of the schema in Python.

Also introduced in this PR is:

- a poetry environment for managing python dependencies.
- a test to test the pydantic data model on the demo data in the repo
- a CI build that builds, lints, and tests the pydantic data model

one modification was made to the JSON Schema to remove special characters from a description of a single field. These characters were causing errors when generating the pydantic model using the code generator.

Speters1086 committed:
add poetry env

Speters1086 committed:
add model

Speters1086 committed:
Update iea43_wra_data_model.schema.json

Speters1086 committed:
add test

Speters1086 committed:
update env

Speters1086 committed:
linting updates

Speters1086 committed:
linting updates

Speters1086 committed:
Update pydantic_model_creation_notes.md

Speters1086 committed:
Update load_demo_schema_to_model.ipynb

Speters1086 committed:
Create python_model_build_lint_test.yml

Speters1086 committed:
Delete requirements.txt

Speters1086 committed:
Update pyproject.toml

Speters1086 committed:
update env for build

Speters1086 committed:
rebase of dev and update model for new fields

stephenholleran commented:
Hi @Speters1086, 

This looks great! Thanks for contributing. I'm still getting my head around all the files but from the outset it looks good with your good documentation.

I went to use the Notebook but got an error.
![image](https://github.com/user-attachments/assets/8aa2e62c-5f33-4c42-bad4-44672581229b)

This looks like that the 'src' folder is at the same level as the 'tools' folder and not a child of where the Notebook is located.

@abohara what do you think of this? You have more coding experience than me so would be great to get your opinion.

Speters1086 commented:
@stephenholleran, this was a bit of a pain on my end. In the end my setting.json in .vscode looked like this to and everything just worked. 

```json
{
    "terminal.integrated.env.osx": {
      "PYTHONPATH": "${workspaceFolder}/.",
    },
    "terminal.integrated.env.linux": {
      "PYTHONPATH": "${workspaceFolder}/.",
    },
    "terminal.integrated.env.windows": {
      "PYTHONPATH": "${workspaceFolder}/.",
    },
    "python.envFile": "${workspaceFolder}/.env",
    "jupyter.notebookFileRoot": "${workspaceFolder}/.",
    "python.autoComplete.extraPaths": [
        "${workspaceFolder}/."
    ],
    "python.pythonPath": ".venv/Scripts/python.exe"
  }
```

jonssonchristian commented on .github/workflows/python_model_build_lint_test.yml:
```diff
@@ -0,0 +1,68 @@
+on:
+  push:
+    branches:
+      - main
```
Should this also include the `dev` branch?

Speters1086 replied:
Good point will update

jonssonchristian commented on .github/workflows/python_model_build_lint_test.yml:
```diff
@@ -0,0 +1,68 @@
+      - main
+    paths-ignore:
+      - '**.md'
+  pull_request:
```
It looks like odd syntax with a key without a value. Is this just the syntax to say all pull requests and all manual triggers?

Speters1086 replied:
I think so. I just copied this from another project... i will review.

jonssonchristian commented on .github/workflows/python_model_build_lint_test.yml:
```diff
@@ -0,0 +1,68 @@
+    branches:
+      - main
+    paths-ignore:
+      - '**.md'
```
Are the markdowns excluded because they are automatically generated?

Speters1086 replied:
We normally add this to exclude builds that are markdown changes only. (no code changes)

jonssonchristian replied:
That makes sense. I had misunderstood what this setting means. Thanks for clarifying.

jonssonchristian commented on .github/workflows/python_model_build_lint_test.yml:
```diff
@@ -0,0 +1,68 @@
+  workflow_dispatch:
+
+env:
+  PYTHON_VERSION: '3.11'
```
If desired to check multiple Python versions, it would be possible to do something like this.

```yaml
jobs:
  test:
    strategy:
      matrix:
        python-version: ["3.11", "3.12", "3.13"]
```

Speters1086 replied:
I will leave this for a separate PR as other testing strategies may want to be implemented.

jonssonchristian commented on pyproject.toml:
```diff
@@ -0,0 +1,68 @@
+[tool.poe.tasks.fmt]
+help = "Format code"
+sequence = [
+    { cmd = "poetry run ruff format src tests" },
```
When I run `ruff check` and `ruff format` in a pipeline without poetry I have to use the following to trigger an error on fix. I am not sure if poetry takes care of that.

```
ruff check --exit-non-zero-on-fix
ruff format --check
```

Speters1086 replied:
I think a couple of things are going on here. The pipeline is using `poe` tasks and is using the `Lint` task:

```python
[tool.poe.tasks]
[tool.poe.tasks.lint]
help = "Lint"
sequence = [
    { cmd = "poetry run ruff check src tests" },
    { cmd = "poetry run mypy src tests" },
]
```

I think the default behaviour of `ruff check <directory>` is to error if there is a linting issue. you can see in the `format` task that both `ruff check --fix` and `ruff format` are run. These fix linting and format the code. There is no check for formatting at the moment. The intention is that the user would format and lint their code. I can see how this could be missed, so I can add the format check to the step.

jonssonchristian replied:
That makes sense. I had to use `ruff check --exit-non-zero-on-fix` elsewhere since I had set the default behaviour of `ruff check` to automatically apply safe fixes, and then it would return an error only if there are issues that cannot be automatically fixed (and in the pipeline run the files cannot be changed).

It might be useful to set up pre-commit hooks for formatting, linting and type checking also, but that would of course definitely be something for a separate PR.

jonssonchristian commented on tests/test_iea43_wra_data_model.py:
```diff
@@ -0,0 +1,23 @@
+def test_iea43_wra_data_model(file_path: str) -> None:
+    with open(file_path) as f:
+        data = json.load(f)
+    wra_data_model = IeaWindResourceAssessmentDataModel(**data)
```
Pydantic models also have the method `model_validate_json(...)` to parse JSON data directory. But here is does not really matter of course, since performance is not an issue.

Speters1086 replied:
`model_validate_json` takes a json string as input. I have updated the test to use `model_validate` which will take a dict.

jonssonchristian replied:
Yes, to use `model_validate_json` you would just parse the file content as a string and then pass straight as the argument rather than using the Python JSON parser. For large workloads, that would be more efficient (as the Pydantic JSON parser is part of Pydantic Core written in Rust), but here that does of course not really matter. And for debugging it might be useful to have the `dict`.

jonssonchristian commented on tests/test_iea43_wra_data_model.py:
```diff
@@ -0,0 +1,23 @@
+
+
+@pytest.mark.parametrize("file_path", get_demo_data_files())
+def test_iea43_wra_data_model(file_path: str) -> None:
```
I think it would also be nice to add a test to validate the examples against the JSON Schema, but that is perhaps not quite within the scope of this PR and can always be done afterwards.

Speters1086 replied:
I think this is outside the scope of this PR. But think it would be nice to add in the future.

jonssonchristian replied:
That makes sense. I can volunteer to make a small addition to add that once this is merged.

jonssonchristian commented on src/iea43_wra_data_model.py:
```diff
@@ -0,0 +1,1381 @@
+class MastSectionGeometryItem(BaseModelWithHash):
+    model_config = ConfigDict(extra="allow")
+    # TODO: add validation for uuid and maybe add constructor to generate uuid
+    uuid: str | None = Field(
```
You can validate a UUID like this.

```python
import uuid as uuid_

uuid: uuid_.UUID | None = Field(...)
```

I think you could do the generation like this (presuming UUID4 is ok).

```python
Field(
    default_factory=uuid.uuid4
    ...
)
```

Speters1086 replied:
Thanks for this. Currently, the JSON schema specifies that the UUIDs are nullable. This is strange to me. If they were not nullable, then adding the validator to generate them would make sense.

jonssonchristian replied:
I think the UUID values are nullable mainly for backwards compatibility reasons. Some of those fields were not present in earlier versions of the schema. For some use cases, it is also not necessary to have universal unique identifiers (or any identifiers) for component objects. In any future major version upgrade with other breaking changes, I would argue to make the UUID fields required and not nullable, to facilitate easier data integration in a wider context, where you might want to reference these individual objects.

As it stands now, it seems like a safer approach not to generate the UUIDs if not included. Then if the data originator does not include UUIDs and the document is processed by different intermediaries with the Pydantic model, the same object could be given different UUIDs.

jonssonchristian commented on src/iea43_wra_data_model.py:
```diff
@@ -0,0 +1,1381 @@
+            if isinstance(value, BaseModel):
+                continue
+            elif isinstance(value, set):
+                continue
```
If it skips fields that are sub-models and sets, could it hash two model instances with different values for those the same?

Speters1086 replied:
Good point. I had done this to get it working and never came back to it. I will have another look.

jonssonchristian commented on src/iea43_wra_data_model.py:
```diff
@@ -0,0 +1,1381 @@
+        None,
+        description="The reference bin for the calibration.",
+        examples=[5.93],
+        title="Reference Bin",
```
The title of a field is automatically generated with this convension (capitalised with spaces) when not specified explicitly, so it would be possible to drop title specifications except for ones that break that convention.

Speters1086 replied:
Not going to change this. I think in terms of the model, this was autogenerated based on json schema.

jonssonchristian commented:
Resolves #244

jonssonchristian commented:
I think this is a great contribution that will provide great ergonomics for working with the data model in a Python environment.

stephenholleran commented:
Thanks @jonssonchristian for taking a look. Some of your comments are way over my understanding so appreciate someone else taking a look.

Speters1086 commented:
Thanks @jonssonchristian for the review. I am on holiday and will get back to this PR when I am back. 

@stephenholleran how close is the floating lidar branch to being merged into dev/main, if not done already? 

I will hold this PR until that branch is committed to dev so I can update the pydantic model with any changes. 

With this PR the test data in the repo is used to test the model. Is the test data exhaustive? I had thought about adding test coverage to give an idea about this. @jonssonchristian thoughts?

jonssonchristian commented:
I'm just about to leave for holidays and will return the week after next. I'll have a look and share thoughts when I get back.

stephenholleran commented:
Hi @Speters1086,

> @stephenholleran how close is the floating lidar branch to being merged into dev/main, if not done already?

The floating lidar file format #271 is pretty much done. Just waiting on at least another person to review it. Aim is to merge it into the dev branch on the call in 2 weeks, 1st May.

There is another branch #282 which contains additional sensor_types and measurement_types falling out from the floating lidar discussions. This PR is up to date with all the decisions currently made but there are a few more items to discuss. Things about wave spectra or wave heights/periods that may or may not required any changes. These are the last few points in today's call that we didn't get to https://github.com/IEA-Task-43/digital_wra_data_standard/discussions/268#discussioncomment-12869659
These could take a few weeks yet.

Thanks again!

abohara commented:
Hi @stephenholleran , @Speters1086 , @jonssonchristian 

Thanks for all your work on this. I realize I am late to this conversation, so please just take these as suggestions. I very much value having Pydantic models that can be leveraged in python code for type checking. However, I do have some arguments against including it in the current form (i.e. in the root folder ) prominently in this repository .

1. This repository is primarily a specification expressed via JsonSchema. We have a JSON example that we keep in sync as well as postgres schema and ipython examples. For every change to the core JSONSchema , we need to update these additional documents to keep them in sync. So, Pydantic models add another long term maintenance responsibilities to any future updates ( though I do realize it is auto generated by github actions ).  We currently have the documentation generation and the form generator that were added in similar vein but we have add issues troubleshooting and keeping them working ( we probably should remove the form generation ). So, the long term maintenance of supporting Pydantic models I think is a concern , especially given relatively recent changes to v2 that did break backward compatibility ( I believe ) and a likely change to v3 in the future .

2. Since, the Pydantic model code is auto generated, would updating the 'Readme' to inform how a user could leverage the 'JSONSchema' to generate & Pydantic models , be a lighterweight but sufficient alternative to help a user generate Pydantic code ?

3. If we do include it , could it put it within the tools section like we do for the postgres schema generation ? This way we would not detract from the core specification available as the JSON schema ? It would also serve as a useful example if another use wanted to add a similar Typescript model ?

4. Or maybe even in its own repository with a prominent link to it ?

I think the Pydantic models are a great addition , not to mention they are much more readable compared to a JSONSchema. I am mostly concerned about the long term maintainability and the ergonomics  of including it as a core part of the repo. Again apologies for jumping in late here, but @stephenholleran hope these help somewhat with the discussion.

Speters1086 commented:
@abohara, you make a valid point. I struggled with where to put this a bit, and hadn't thought to put it in the tools folder. I think this makes more sense, so I will address the other PR comments and then refactor the structure. I will also update the `README.md` to include a section explaining the Pydantic model. If we captured code coverage, this would also be displayed in the section, along with the build status for the model tests. 

I also agree that there is an overhead to maintaining the Pydantic model, however, my view is that the benefits of ease of use test testability outweigh the overhead cost, but happy to take other views. You are correct that a code generator is used to create the model, but it is not automatically done. The current process is described in `src/pydantic_model_creation_notes.md`. 

The other aspect of this is how settled this standard is. From the change history, it looks like the pace of change is relatively low, and thus, the overhead will be minor.

Speters1086 committed:
Update python_model_build_lint_test.yml

Speters1086 committed:
Update pyproject.toml

Speters1086 committed:
Update iea43_wra_data_model.py

Speters1086 committed:
Update test_iea43_wra_data_model.py

Speters1086 committed:
update env

Speters1086 committed:
restructure repo and update notebooks

Speters1086 committed:
Update README.md

Speters1086 committed:
update env

Speters1086 commented:
I have made some changes based on the PR comments and would like to ask for additional review. 

I have rebased my changes on the latest version of the `dev` branch. 

@jonssonchristian, I have addressed most of your PR comments. Please have a look and let me know if you still have any questions. 

@abohara, I have restructured the repo slightly to pull out the PostgreSQL and Pydantic Models into their own folder and changed the `tools` directory to `examples`. This makes sense as the other tools, such as the `app` were already in their own directory.  I have also updated the `README` to explain the difference between `Tools` and `Examples`. 

I have also updated the example notebooks to work with the poetry environment provided and fixed some minor bugs so they will run though to completion. 

@stephenholleran the `2023 Workshop - Floating Lidar Demo - Daniel Nuno.ipynb` is dependent on data that is not in the repo. Should this example be removed?

jonssonchristian commented:
> With this PR the test data in the repo is used to test the model. Is the test data exhaustive? I had thought about adding test coverage to give an idea about this. @jonssonchristian thoughts?

I think it would be useful to add test coverage.

All that should be needed is to add `pytest-cov` to the external package requirements and then to include the below (or similar) in the `pyproject.toml` file or as CLI arguments in the pytest step. I am however not sure if there is a poetry-specific way of doing it.

```toml
[tool.pytest.ini_options]
addopts = [
    "--cov=wra_data_model_pydantic",
    "--cov-report=html",
    "--cov-report=xml",
    "--cov-report=term-missing",
]
```

jonssonchristian reviewed (approved):
Hi @Speters1086,

Thank you for the clarifications and updates. It looks good to me. In the following are some general comments and ideas, which may or may not be relevant, and some are certainly outside the scope of this PR and more as possible future developments.

- When we represent JSON arrays as sets, are we sure that users never rely on the order of the elements? I imagine the order will not be significant, more than perhaps for convenience when displaying the items. If someone parses JSON data into the Pydantic model, edits some data, and writes back to JSON, all the arrays may have random order. If there are duplicate items, the duplicates will simply be dropped without raising an error or warning that there is an issue with the input data. The JSON Schema arrays have `"uniqueItems": true`, so validation would fail if the items were not all unique.

- I agree that the Pydantic model increases maintenance effort, but should be worthwhile given how many users are working in a Python environment and how much it can benefit them. From my perspective, the most important thing would be to add automatic checks to ensure that the JSON Schema and Pydantic model fall out of sync. The tests on the examples already cover that, and more checks can probably be added in due course. It is very easy to miss something when aligning different content (even making sure all examples are updated correctly when updating the schema), and built-in checks can be really helpful.

- I think it would be useful to publish the Pydantic model in a pip package so that users can install it very easily into their Python environment and control the version dependency, rather than fetching the files from GitHub.

- There may be some advantages of separating all the Python-related content into a separate directory, also the `pyproject.toml`, poetry related files and tests. That might make it easier to navigate what files are related to what. The [EYA DEF repo](https://github.com/IEC-61400-15/eya_def/tree/main) is currently structured like that, where `eya_def_tools` is the Python library.

Speters1086 committed:
Update iea43_wra_data_model.py

Speters1086 committed:
Update README.md

Speters1086 commented:
@jonssonchristian, Thank you for your help with this and sorry it has taken me so long to get back to this. Your first point above is a very good one and something I had meant to come back to. I have updated the Pydantic model to no longer use `sets` and to validate the uniqueness of lists instead. As for the other points, I agree with them all, but think they should be addressed in separate PRs. 

@stephenholleran, I know there is an upcoming release of the schema. I have rebased this branch of dev so it should be up to date. I am not sure about the protocol, as this is my first contribution here. Should I merge this into the dev branch?

stephenholleran commented:
Hi @Speters1086,

Thanks for your work on this.

Sorry I haven't had much time to spend on this. I have been focused on getting the additional floating lidar parameters over the line for the next release. I had hoped to have completed that last Friday but there are some reviews that I still need to deal with. We decided on the call last Thursday to get the release out and then focus on this Pydantic work you have been doing and do another minor release (if it needs a release as it is not impacting the Schema). With WindEurope this week and WESC next week it is going to be challenging.

I can merge it once I get my head around it. I would like to go back to Amit's comments and understand those better too as it will most likely be me that will be trying to keep it updated.

Cheers,

olivia-bentley commented on wra_data_model_pydantic/iea43_wra_data_model.py:
```diff
@@ -0,0 +1,1471 @@
+        title="Organisation",
+    )
+    date_created: date = Field(
+        ...,
```
renamed from date?

olivia-bentley commented on wra_data_model_pydantic/iea43_wra_data_model.py:
```diff
@@ -0,0 +1,1471 @@
+        description="The uncertainty of the logger acquisition system as a percentage. This is also referred to as the "
+        "uncertainty of the data acquisition system within IEC 61400-12-1. ",
+        examples=["0.1"],
+        title="Logger Acquisition Uncertainty [%]",
```
why is the example "0.1" this applied to the json schema as that's where it comes from

olivia-bentley commented on wra_data_model_pydantic/iea43_wra_data_model.py:
```diff
@@ -0,0 +1,1471 @@
+    @field_validator("classification")
+    def validate_classification(cls, value: str) -> str:  # noqa: N805
+        pattern = re.compile(r"^[0-9]\.[0-9][ABCDS]$")
+        if value and not pattern.match(value):
```
the validator pattern here is slightly different

olivia-bentley committed:
addition of missing descriptions, updated regex pattern for classification and notebook fixes

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/IEA-Task-43/digital_wra_data_standard?pullRequest=281) <br/>All committers have signed the CLA.

olivia-bentley commented:
Hi @Speters1086, I've been reviewing this branch on Stephen's behalf and made a few small changes in the last commit. Please take a look and let me know if you have any concerns or would like to discuss any of the changes. Thanks for all your work on this I think incorporating pydantic here is a very valuable addition.

Speters1086 commented:
@olivia-bentley, sorry, this has gone off the boil for me. I can get back to this in a bit. it looks like this branch is quite a bit behind the dev branch, so it probably needs to be updated before this can be reviewed again. I can try to prioritise this in the next couple of weeks. 

The main question is, are we happy with the structure the way it is or should the repo be refactored? I am not pressed about this, but it will mean changing some build steps.

olivia-bentley commented:
@Speters1086 myself and Stephen would be happy to take on the work to bring it up to date with dev and think about the structure. It will be great to be able to bring the work into the library with credit to you for adding this.

Speters1086 commented:
@olivia-bentley, no worries... happy for you to get this over the line.

olivia-bentley committed:
tools folder restructure and notebook path/bug fixes
